src/day14.scala

```diff
@@ -0,0 +1,112 @@
+type CharPair = (Char, Char)
+type InsertionRules = Map[(Char, Char), Char]
+type Frequencies = Map[Char, Long]
+
+def part1(input: String): Long =
+  val (initialPolymer, insertionRules) = parseInput(input)
+  val finalPolymer = (0 until 10).foldLeft(initialPolymer)((polymer, _) => applyRule(polymer, insertionRules))
+  val frequencies: Frequencies = finalPolymer.groupMapReduce(identity)(_ => 1L)(_ + _)
+  val max = frequencies.maxBy((char, count) => count)._2
+  val min = frequencies.minBy((char, count) => count)._2
+  max - min
+
+def parseInput(input: String): (Polymer, InsertionRules) =
+  val sections = input.split("\n\n")
+  val initialPolymer = sections(0).toList
+  val insertionRules = sections(1).linesIterator.map(parseRule).toMap
+  (initialPolymer, insertionRules)
+
+def parseRule(line: String): (CharPair, Char) =
+  line match
+    case s"$pairStr -> $inserStr" => (pairStr(0), pairStr(1)) -> inserStr(0)
+    case _                        => throw new Exception(s"Cannot parse '$line' as an insertion rule")
+
+def applyRule(polymer: Polymer, rules: InsertionRules): Polymer =
+  val pairs = polymer.zip(polymer.tail)
+  val insertionsAndSeconds: List[List[Char]] =
+    for pair <- pairs
+      yield rules(pair) :: pair._2 :: Nil
+  polymer.head :: insertionsAndSeconds.flatten
+
+def part2(input: String): Long =
+  val (initialPolymer, insertionRules) = parseInput(input)
+
+  /* S(ab...f, n) :=
+   *   frequencies of all letters *except the first one* in the expansion of "ab...f" after n iterations
+   *
+   * Frequencies are multisets. We use + and ∑ to denote their sum.
+   *
+   * For any string longer than 2 chars, we have
+   *   S(x₁x₂x₃...xₚ, n) = ∑(S(xᵢxⱼ, n)) for i = 0 to n-1 and j = i+1
+   * because each initial pair expands independently of the others. Each initla
+   * char is counted exactly once in the final frequencies because it is
+   * counted as part of the expansion of the pair on its left, and not the
+   * expansion of the pair on its right (we always exclude the first char).
+   *
+   * As particular case of the above, for a string of 3 chars xzy, we have
+   *   S(xzy, n) = S(xz, n) + S(zy, n)
+   *
+   * For strings of length 2, we have two cases: n = 0 and n > 0.
+   *
+   * Base case: a pair 'xy', and n = 0
+   *   S(xy, 0) = {y -> 1} for all x, y
+   *
+   * Inductive case: a pair 'xy', and n > 0
+   *   S(xy, n)
+   *     = S(xzy, n-1) where z is the insertion char for the pair 'xy' (by definition)
+   *     = S(xz, n-1) + S(zy, n-1) -- the particular case of 3 chars above
+   *
+   * And that means we can iteratively construct S(xy, n) for all pairs 'xy'
+   * and for n ranging from 0 to 40.
+   */
+
+  // S : (charPair, n) -> frequencies of everything but the first char after n iterations from charPair
+  val S = mutable.Map.empty[(CharPair, Int), Frequencies]
+
+  // Base case: S(xy, 0) = {y -> 1} for all x, y
+  for (pair, insert) <- insertionRules do
+    S((pair, 0)) = Map(pair._2 -> 1L)
```

Comment:
```suggestion
    S((pair, 0)) = Map(pair(1) -> 1L)
```

(I like the homogeneity with collections)